lonboard/_geoarrow/movingpandas_interop.py

```diff
@@ -0,0 +1,199 @@
+        coords[start_offset:end_offset, 1] = shapely.get_y(traj.df.geometry)
+        coords[start_offset:end_offset, 2] = shapely.get_z(traj.df.geometry)
+
+        geom_col_name = traj.get_geom_col()
+        df_attr = traj.df.drop(columns=[geom_col_name])
+
+        attr_table = pa.Table.from_pandas(
+            traj.df, schema=attr_schema, preserve_index=False
+        )
+        attr_tables.append(attr_table)
+
+    attr_table = pa.concat_tables(attr_tables, promote_options="none")
+    attr_table = Table.from_arrow(attr_table)
+
+    offsets = Array.from_numpy(offsets)
+
+    nested_attr_table = apply_offsets_to_table(attr_table, offsets=offsets)
+
+    if np.alltrue(np.isnan(coords[:, 2])):
+        coord_list_size = 2
+        # Cast to 2D coords
+        coords = coords[:, :2]
+    else:
+        assert not np.any(
+            np.isnan(coords[:, 2])
+        ), "Mixed 2D and 3D coordinates not currently supported"
+        coord_list_size = 3
+
+    coords_arr = Array.from_numpy(coords.ravel("C"))
+    coords_fixed_size_list = fixed_size_list_array(coords_arr, coord_list_size)
+    linestrings_arr = list_array(offsets, coords_fixed_size_list)
+
+    extension_metadata: Dict[str, str] = {"ARROW:extension:name": "geoarrow.linestring"}
+    if crs_json is not None:
+        extension_metadata["ARROW:extension:metadata"] = json.dumps({"crs": crs_json})
+
+    linestrings_field = Field(
+        "geometry",
+        linestrings_arr.type,
+        nullable=True,
+        metadata=extension_metadata,
+    )
+
+    timestamp_values = Array.from_numpy(timestamps).cast(time_arrow_dtype)
+    timestamp_arr = list_array(offsets, timestamp_values)
+    timestamp_col = ChunkedArray([timestamp_arr])
+
+    table = nested_attr_table.append_column(
+        linestrings_field, ChunkedArray([linestrings_arr])
+    )
+    return table, timestamp_col
+
+
+def infer_time_unit(dtype: np.dtype) -> Tuple[Literal["s", "ms", "us", "ns"], DataType]:
+    """Infer an arrow time unit from the numpy data type
+
+    Raises:
+        ValueError: If not a known numpy datetime dtype
+    """
+
+    if dtype.name == "datetime64[s]":
+        code = "s"
+        return code, DataType.timestamp(code)
+
+    if dtype.name == "datetime64[ms]":
+        code = "ms"
+        return code, DataType.timestamp(code)
+
+    if dtype.name == "datetime64[us]":
+        code = "us"
+        return code, DataType.timestamp(code)
+
+    if dtype.name == "datetime64[ns]":
+        code = "ns"
+        return code, DataType.timestamp(code)
+
+    raise ValueError(f"Unexpected datetime type: {dtype}")
+
+
+def apply_offsets_to_table(table: Table, offsets: Array) -> Table:
+    batch = table.combine_chunks().to_batches()[0]
+
+    new_fields = []
+    new_arrays = []
+
+    for field_idx in range(batch.num_columns):
+        field = batch.schema.field(field_idx)
+        new_field = field.with_type(DataType.list(field))
```

Comment:
This relies on arro3's `list(..)` to convert the pyarrow field through the arrow pycapsule interface, and then pyarrow's `with_type()` convert again in the other direction? 
Nice ;)

Reply:
Ah, no, I see that you converted the pyarrow Table to am arro3 one (`attr_table = Table.from_arrow(attr_table)`) before calling this function ;)

Reply:
Here `batch` is actually an `arro3.RecordBatch`, the conversion back to arro3 happened here:
https://github.com/developmentseed/lonboard/blob/1ced62191ade5a110d2eef2df4524c5a5a49d740/lonboard/_geoarrow/movingpandas_interop.py#L116

But yes, these lines should work with pyarrow input as well!

Reply:
> Ah, no, I see that you converted the pyarrow Table to am arro3 one (`attr_table = Table.from_arrow(attr_table)`) before calling this function ;)

Right. I could've left it as a pyarrow table; the main difference is mostly for type hinting. It's nice to get IDE completions and type checking and pyarrow doesn't have that yet.